src/main/java/org/passay/RuleResultMetadata.java

```diff
@@ -0,0 +1,89 @@
+/* See LICENSE for licensing and NOTICE for copyright. */
+package org.passay;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * Describes metadata relevant to the result of rule validation.
+ *
+ * @author  Middleware Services
+ */
+public class RuleResultMetadata
+{
+
+  /** Additional metadata that provide information about the rule. */
+  protected final Map<String, Object> metadata;
+
+
+  /**
+   * Creates a new rule result metadata.
+   *
+   * @param  data  metadata.
+   */
+  public RuleResultMetadata(final Map<String, Object> data)
+  {
+    if (data == null) {
+      metadata = new LinkedHashMap<>();
+    } else {
+      metadata = new LinkedHashMap<>(data);
+    }
+  }
+
+
+  /**
+   * Puts the supplied key and value into this metadata object.
+   *
+   * @param  key  of the metadata.
+   * @param  value  of the metadata.
+   */
+  public void put(final String key, final Object value)
+  {
+    metadata.put(key, value);
+  }
+
+
+  /**
+   * Puts all the supplied metadata into this metadata object.
+   *
+   * @param  data  to add.
+   */
+  public void putAll(final Map<String, Object> data)
+  {
+    metadata.putAll(data);
+  }
+
+
+  /**
+   * Returns the metadata value for the supplied key.
+   *
+   * @param  <T>  type of value to return
+   * @param  key  to the metadata to retrieve.
+   * @param  type  to cast the metadata value to.
+   *
+   * @return  metadata value.
+   */
+  public <T> T get(final String key, final Class<T> type)
+  {
+    return type.cast(metadata.get(key));
+  }
+
+
+  /**
+   * Returns an unmodifiable map of all the metadata.
+   *
+   * @return  all metadata.
+   */
+  public Map<String, Object> getAll()
+  {
+    return Collections.unmodifiableMap(metadata);
```

Comment:
Why not simply expose the map for mutations? Since you can mutate the underlying map with the put operations, returning an unmodifiable collection here doesn't provide much if anything. The type-sensitive getter has some value I suppose, but you only have a use case for `Map<String, Integer>` today, so you could just use that signature.

Reply:
Since this class only wraps a `Map` I made as well provide some API value by controlling the mutators. That's the only reason for making it unmodifiable.

If we want to say all metadata is an `Integer`, I can make that change.

Reply:
+1 to change to `Integer`. Then the getter can return `Integer` (or null) without any fancy type checking.